### PR TITLE
fix: address code review findings in detection rules and hook output

### PR DESCRIPTION
**Pull request by chataclaw**
## Summary

- Use nullish coalescing (`??`) instead of logical OR (`||`) in `entropy()` for correct semantics under `noUncheckedIndexedAccess`
- Add negative lookahead to `openai-key` regex (`(?!proj-|ant-)`) to prevent overlapping matches with `openai-project-key` and `anthropic-key`
- Remove unreachable `unique` filter in `user-prompt-submit-hook` (dead code after mask branch exit)
- Consolidate `randomBird()` calls in `block()` to single invocation for consistent emoji in terminal and JSON output
- Remove redundant `randomBird()` from `detectionLines` arguments — emoji is already rendered in the `block()` header line
- Add `gcp-api-key` and `npm-token` detection rules with tests
- Add regression tests for `openai-key` negative lookahead
- Limit file content scanning to first 1 MB and transcript reading to last 64 KB for performance
- Skip binary files by detecting NUL bytes in the first 8 KB
- Use `bytesRead` return value from `fs.readSync` and wrap fd operations in `try/finally`

## Known tradeoffs

- **File scan limit (1 MB):** Secrets beyond the first 1 MB of a file will not be detected. This is an intentional tradeoff — secrets in configuration files, source code, and `.env` files are overwhelmingly located near the beginning. Scanning unbounded file sizes risks OOM and long hook latency.
- **Transcript tail read (64 KB):** Only the last 64 KB of the transcript is read for allow-tag resolution. A single JSONL user message exceeding 64 KB could cause the allow tag to be missed, falling back to the default block behavior (safe side). In practice, user messages are well under this limit.

## Test plan

- [x] `pnpm run ci` passes (typecheck + biome check + 187 tests)
- [ ] Verify block messages display single consistent bird emoji
- [ ] Verify `sk-proj-*` and `sk-ant-*` keys are not matched by `openai-key` rule
- [ ] Verify GCP API key and npm token detection
- [ ] Verify binary files are skipped

### CODE REVIEW

File: src/pre-tool-use-hook.ts [hunk 1]
```diff
@@ -218,15 +245,29 @@ function scanFile(filePath: string, allowTags: Set<string>): void {
...
+    const fd = fs.openSync(filePath, "r");
+    let bytesRead: number;
+    try {
+      bytesRead = fs.readSync(fd, buf, 0, bytesToRead, 0);
+    } finally {
+      fs.closeSync(fd);
+    }
+    const data = buf.subarray(0, bytesRead);
+    // Skip binary files: if NUL byte exists in the first 8KB, assume binary
+    if (data.subarray(0, 8192).includes(0)) return;
```
[the remark this answers is not in the document]
Comment (chataclaw) [reply src/pre-tool-use-hook.ts]: Thank you for the suggestion. We considered this but decided to keep the current behavior (silently allowing binary reads) for the following reasons:

1. **sensitive-canary is a text-based regex scanner.** Running regex patterns against binary data cannot detect secrets embedded in images (e.g., screenshots containing passwords) — that would require OCR, which is out of scope for this tool.
2. **False positives from binary data.** Byte patterns in images, compiled binaries, and other non-text files can accidentally match secret patterns, leading to noise without security benefit.
3. **Claude Code legitimately reads binary files.** The Read tool is used for images in multimodal workflows. Blocking all binary reads would disrupt normal usage and require frequent allow-tag overrides.

The scope of sensitive-canary is intentionally limited to detecting secrets and PII in **text-readable content**. Detecting sensitive information in images would be better handled by a separate OCR-based layer.

File: src/pre-tool-use-hook.ts [hunk 2]
```diff
@@ -218,15 +248,29 @@ function scanFile(filePath: string, allowTags: Set<string>): void {
...
+    content = data.toString("utf8");
```
[the remark this answers is not in the document]
Comment (chataclaw) [reply src/pre-tool-use-hook.ts]: Good point about the bypass risk. Updated the implementation to scan the text prefix before the first NUL byte instead of returning early. This way:

- Secrets in the text portion before NUL → blocked
- Secrets only after NUL → allowed (not readable as text anyway)
- File starting with NUL → allowed (no text to scan)

This preserves the performance benefit of not running regex on raw binary data while closing the bypass.

**Review by chata — APPROVED** [review]
